nixops/resources/aws_vpn_connection.py

```diff
@@ -0,0 +1,146 @@
+        self._state = StateDict(depl, id)
+        self._client = None
+        self.handle_create_vpn_conn = Handler(['region', 'vpnGatewayId', 'customerGatewayId', 'staticRoutesOnly'], handle=self.realize_create_vpn_conn)
+        self.handle_tag_update = Handler(['tags'], after=[self.handle_create_vpn_conn], handle=self.realize_update_tag)
+
+    @classmethod
+    def get_type(cls):
+        return "aws-vpn-connection"
+
+    def show_type(self):
+        s = super(AWSVPNConnectionState, self).show_type()
+        if self._state.get('region', None): s = "{0} [{1}]".format(s, self._state['region'])
+        return s
+
+    @property
+    def resource_id(self):
+        return self._state.get('vpnConnectionId', None)
+
+    def prefix_definition(self, attr):
+        return {('resources', 'awsVPNConnections'): attr}
+
+    def get_defintion_prefix(self):
+        return "resources.awsVPNConnections."
+
+    def connect(self):
+        if self._client: return
+        self._client = nixops.ec2_utils.connect_ec2_boto3(self._state['region'], self.access_key_id)
+
+    def create_after(self, resources, defn):
+        return {r for r in resources if
+                isinstance(r, nixops.resources.vpc_customer_gateway.VPCCustomerGatewayState) or
+                isinstance(r, nixops.resources.aws_vpn_gateway.AWSVPNGatewayState)}
+
+    def create(self, defn, check, allow_reboot, allow_recreate):
+        diff_engine = self.setup_diff_engine(config=defn.config)
+
+        self.access_key_id = defn.config['accessKeyId'] or nixops.ec2_utils.get_access_key_id()
+        if not self.access_key_id:
+            raise Exception("please set 'accessKeyId', $EC2_ACCESS_KEY or $AWS_ACCESS_KEY_ID")
+
+        for handler in diff_engine.plan():
+            handler.handle(allow_recreate)
+
+    def realize_create_vpn_conn(self, allow_recreate):
+       config = self.get_defn()
+
+       if self.state == self.UP:
+            if not allow_recreate:
+                raise Exception("vpn connection {} definition changed and it needs to be recreated"
+                                " use --allow-recreate if you want to create a new one".format(self._state['vpnConnectionId']))
+            self.warn("vpn connection definition changed, recreating ...")
+            self._destroy()
+            self._client = None
+
+       self._state['region'] = config['region']
+       self.connect()
+
+       customer_gtw_id = config['customerGatewayId']
+       if customer_gtw_id.startswith("res-"):
+           res = self.depl.get_typed_resource(customer_gtw_id[4:].split(".")[0], "vpc-customer-gateway")
+           customer_gtw_id = res._state['customerGatewayId']
+
+       vpn_gateway_id = config['vpnGatewayId']
+       if vpn_gateway_id.startswith("res-"):
+           res = self.depl.get_typed_resource(vpn_gateway_id[4:].split(".")[0], "aws-vpn-gateway")
+           vpn_gateway_id = res._state['vpnGatewayId']
```

Comment:
Since this idiom appears so frequently, we should create a helper function for it, e.g.
```
vpn_gateway_id = get_resource('vpnGatewayId', "aws_vpn-gateway")
```

Also, we should check whether `res._state['vpnGatewayId']` is initialized. I'm not sure if that's guaranteed to be the case when using `nixops deploy --include ...` (since some dependencies might be excluded).